File: tree/1_maximum_depth_of_binary_tree_104.cc
#include "binary_tree.h"

class Solution1 {
public:
    int depth = 0, res = 0;

    int maxDepth(TreeNode* root) {
        traverse(root);
        return res;
    }

    void traverse(TreeNode* root) {
        if (nullptr == root) return;

        ++depth;
        if (nullptr == root->left && nullptr == root->right) {
            res = max(res, depth);
        }
        if (root->left) traverse(root->left);
        if (root->right) traverse(root->right);
        --depth;
        return;
    }
};

class Solution2 {
public:
    int maxDepth(TreeNode* root) {
        if (nullptr == root) return 0;
        int res = max(maxDepth(root->left), maxDepth(root->right)) + 1;
        return res;
    }
};

class Solution3 {
public:
    int maxDepth(TreeNode* root) {
        if (!root) return 0;

        queue<TreeNode*> que;
        que.push(root);
        int ans = 0;
        while (!que.empty()) {
            int sz = que.size();
            while (sz > 0) {
                TreeNode *cur = que.front();
                que.pop();
                if (cur->left) que.push(cur->left);
                if (cur->right) que.push(cur->right);
                --sz;
            }
            ++ans;
        }
        return ans;
    }
};

File: tree/1_maximum_depth_of_binary_tree_104.py
import binary_tree

class Solution1:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1

class Solution2:
    def __init__(self):
        self.depth, self.res = 0, 0
    
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        self.traverse(root)
        return res
    
    def traverse(self, root: Optional[TreeNode]) -> None:
        if not root:
            return
        self.depth += 1
        if not root.left and not root.right:
            self.res = max(self.res, self.depth)
        if root.left:
            traverse(root.left)
        if root.right:
            traverse(root.right)
        self.depth -= 1
        return

class Solution3:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        queue = [root]
        depth = 0

        while queue:
            n = len(queue)
            for i in range(n):
                node = queue.pop(0)
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            depth += 1
        return depth
